streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java

```diff
@@ -0,0 +1,1123 @@
+                final Set<Long> keysSecondClient = keysFromInstance(streams2Alpha);
+
+                final List<KeyValue<Long, Long>> committedInputDataAfterFirstUpgrade =
+                    prepareData(15L, 20L, keysFirstClient.toArray(new Long[0]));
+                writeInputData(committedInputDataAfterFirstUpgrade);
+
+                final List<KeyValue<Long, Long>> expectedCommittedResultBeforeFailure =
+                    computeExpectedResult(committedInputDataAfterFirstUpgrade, committedState);
+                verifyCommitted(expectedCommittedResultBeforeFailure);
+                expectedUncommittedResult.addAll(expectedCommittedResultBeforeFailure);
+
+                commitCounterClient2.set(0);
+
+                final Iterator<Long> it = keysSecondClient.iterator();
+                final Long otherKey = it.next();
+                final Long failingKey = it.next();
+
+                final List<KeyValue<Long, Long>> uncommittedInputDataAfterFirstUpgrade =
+                    prepareData(15L, 19L, keysSecondClient.toArray(new Long[0]));
+                uncommittedInputDataAfterFirstUpgrade.addAll(prepareData(19L, 20L, otherKey));
+                writeInputData(uncommittedInputDataAfterFirstUpgrade);
+
+                final Map<Long, Long> uncommittedState = new HashMap<>(committedState);
+                expectedUncommittedResult.addAll(
+                    computeExpectedResult(uncommittedInputDataAfterFirstUpgrade, uncommittedState)
+                );
+                verifyUncommitted(expectedUncommittedResult);
+
+                stateTransitions1.clear();
+                stateTransitions2.clear();
+                commitCounterClient1.set(0);
+                commitErrorInjectedClient2.set(true);
+
+                final List<KeyValue<Long, Long>> dataFailingKey = prepareData(19L, 20L, failingKey);
+                uncommittedInputDataAfterFirstUpgrade.addAll(dataFailingKey);
+                writeInputData(dataFailingKey);
+
+                expectedUncommittedResult.addAll(
+                    computeExpectedResult(dataFailingKey, uncommittedState)
+                );
+                verifyUncommitted(expectedUncommittedResult);
+
+                waitForStateTransition(stateTransitions1, SINGLE_REBALANCE, stateTransitions2, CRASH);
+
+                commitErrorInjectedClient2.set(false);
+                stateTransitions2.clear();
+                streams2Alpha.close();
+                waitForStateTransition(stateTransitions2, CLOSE_CRASHED);
+
+                final List<KeyValue<Long, Long>> expectedCommittedResultAfterFailure =
+                    computeExpectedResult(uncommittedInputDataAfterFirstUpgrade, committedState);
+                verifyCommitted(expectedCommittedResultAfterFailure);
+                expectedUncommittedResult.addAll(expectedCommittedResultAfterFailure);
+            }
+
+            // 7. only for crash case:
```

Comment:
This is the second part of the "mixed mode" test: client1 is on eos-beta and client2 is on eoa-alpha. In the first part above, we crashed the second client and restarted it in eos-alpha mode. In this second part, we crash client1 and restart it in eos-beta mode.

The actual upgrade continues in the next phase.